### DISCUSSION TITLE
Support RFC 4191 Route Information (Option 24) for Thread

### DISCUSSION THREAD

**Discussion by agners** [home-assistant/operating-system#2333] [Ideas]
Currently Home Assistant OS does not have `CONFIG_IPV6_ROUTER_PREF` or `CONFIG_IPV6_ROUTE_INFO` set. 

These options add support for ICMPv6 Router Advertisements with Route Information Option (Type 24). This type of Router Information are sent by Thread Border Routers.

Note: It seems that the `otbr-agent` processes such advertisements on its own as well:
```
otbr-agent[282]: 06:21:46.999 [I] BorderRouter--: Processing RIO (fd84:da21:5e08:1::/64, 1800 seconds)
otbr-agent[282]: 06:21:46.999 [I] NetDataPublshr: Publishing ExternalRoute fd84:da21:5e08:1::/64
otbr-agent[282]: 06:21:47.000 [I] NetDataPublshr: ExternalRoute fd84:da21:5e08:1::/64 (state:Added) in netdata - total:1, preferred:0, desired:10
```

~~This means that a Home Assistant OS installation which runs a OTBR actually learns about routes to other Thread border routers. However, Home Assistant OS installations without a `otbr-agent` running do not learn the routes to other BRs and therefore can not communicate with Thread devices connected to the Thread network behind those BRs.~~ (see below

See also [this discussion on Discord](https://discord.com/channels/330944238910963714/664014360083038228/1069652144833167470).

**Comment by agners** [comment]
Further investigations actually showed that the route information of another Thread border router even got to be processed when no OTBR was running (and the kernel configurations were disabled). It turns out that the NetworkManager daemon processes IPv6 Neighbor Discovery packets as well. In particular the code in [src/core/ndisc/nm-lndp-ndisc.c#L240-265](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.34.0/src/core/ndisc/nm-lndp-ndisc.c#L240-265) processes RFC 4191 Route Information packets.

So in other words, the kernel configurations are not really necessary for this to work.

Also, the kernel configuration by default sets the max prefix len to 0, which essentially disables the feature. A sysctl configuration to set the `accept_ra_rt_info_max_plen` configuration would be needed. There is an `all` interface configuration, but its unclear if that really works and if it is a good idea to set it.

**Reply by pvizeli** [reply]
If that is enough what NM does, we should keep it like it is. NM knows which interface he can touch and which he should not touch. It feel more safe to me on a docker env

**Reply by agners** [reply]
Yeah agree, I don't see a good reason why we should enable it in the kernel right now, especially since it seems that the NetworkManager disables the kernel feature anyways.

**Comment by Jc2k** [comment]
I am still working through something on my end, but I wanted to post the current situation as I was the trigger for this issue existing.

I am currently helping a HAOS user who has 5 border routers. They are all Apple (mixture of ATV and HomePod). mDNS discovery is working, but user see's a timeout when trying to pair their Eve accessories. Based on the users who I have helped so far, this usually means a problem with IPv6 connectivity. 

My baseline environment is similar to theirs - I have a single HomePod border router, and I have functional Eve devices. I am using the official HA container, but I am not running HAOS yet (my yellow just arrived, though).

My first step is to confirm the HA container sees the icmp6 packet with the Option 24 RI, using tcpdump. This lets me rule out things like VLAN's or strange WiFi repeaters breaking icmp6. The container does see the RI:

```
15:57:57.844450 a8:51:ab:90:6e:d0 > 33:33:00:00:00:01, ethertype IPv6 (0x86dd), length 142: (flowlabel 0x30c00, hlim 255, next-header ICMPv6 (58) payload length: 88) fe80::c24:8ac1:ca8d:c813 > ff02::1: [icmp6 sum ok] ICMP6, router advertisement, length 88
    hop limit 0, Flags [none], pref medium, router lifetime 0s, reachable time 0ms, retrans timer 0ms
      source link-address option (1), length 8 (1): a8:51:ab:90:6e:d0
        0x0000:  a851 ab90 6ed0
      prefix info option (3), length 32 (4): fdbe:e98b:173f:4f12::/64, Flags [onlink, auto], valid time 1800s, pref. time 1800s
        0x0000:  40c0 0000 0708 0000 0708 0000 0000 fdbe
        0x0010:  e98b 173f 4f12 0000 0000 0000 0000
      route info option (24), length 16 (2):  fdcd:7fc0:8284::/64, pref=medium, lifetime=1800s
        0x0000:  4000 0000 0708 fdcd 7fc0 8284 0000
      route info option (24), length 16 (2):  fd21:1896:ac15::/64, pref=medium, lifetime=1800s
        0x0000:  4000 0000 0708 fd21 1896 ac15 0000
```

However at first glance, there was no route information being added to the kernels route table.

On other systems the solution has been to set `accept_ra_rt_info_max_plen` to 128. The kernel then processes the RI and automatically adds the routes with `proto ra`. It has been proposed by SUSE that this be this default for some time, but this has not yet happened. Setting the sysctl is also mentioned in the OpenThread docs.

We tried to do this here, but it was not possible because `CONFIG_IPV6_ROUTE_INFO` is not enabled on HAOS. For reference, whilst I am not advocating for it to be turned on in particular, its worth mentioning it is enabled by default on Ubuntu, Debian, Arch and the Raspberry Pi kernel.

With @agners help, we realised that something in HAOS was processing the route info that wasn't the kernel. As above, this turned out to be NM.

User had ipv6 turned off in HA, which i didn't notice because link local addressing was working (this is expected since https://github.com/home-assistant/supervisor/pull/3829). When we turned it on routes apper in `ip -6 route`, however they look wrong:

```
fd21:1896:ac15::/64  metric 100 
fdbe:e98b:173f:4f12::/64 dev enp0s3  metric 100 
fdcd:7fc0:8284::/64  metric 100 
```

Note that the thread ULA would normally have a via. For example, here is the equivalent entries from my environment. I am using the same HA container to generate this table, and the same Apple hardware, so I would expect to see a route in his environment like this but I don't.

```
fd59:86c6:e5a5::/64 via fe80::cac:1b0:a113:3f18 dev vlan101  metric 1024  expires 0sec
```

I am wondering if the busybox version of ip is to blame here - it is worth calling out that the same route table entry on a Debian box with "full" iproute2 looks like this:

```
fd59:86c6:e5a5::/64 via fe80::cac:1b0:a113:3f18 dev vlan101 proto ra metric 1024 expires 1764sec pref medium
```

So either NM isn't adding correct routes, or the default alpine `ip` command is struggling to show them for this user.

I wasn't expecting this to work, because the route table looks "wrong" to me, but I did try `ip route get` and this makes it look like the routes are there but just not visible with the alpine version of `ip`:

```
# ip route get fdcd:7fc0:8284:0:531a:356a:f5dd:8846
fdcd:7fc0:8284:0:531a:356a:f5dd:8846 via fe80::80f:73cb:ae8:b858 dev enp0s3  src fdbe:e98b:173f:4f12:1763:4bc6:8977:84a9  metric 100
```

So, working theory was `ip` from busybox was insufficient. We switched to iproute2 and now we can see:

```
fd21:1896:ac15::/64 proto ra metric 100 pref medium
    nexthop via fe80::46c:2041:d140:dae0 dev enp0s3 weight 1 
    nexthop via fe80::c24:8ac1:ca8d:c813 dev enp0s3 weight 1 
    nexthop via fe80::184b:fec6:8d00:a266 dev enp0s3 weight 1 
    nexthop via fe80::80f:73cb:ae8:b858 dev enp0s3 weight 1 
    nexthop via fe80::14a2:7660:694b:73ce dev enp0s3 weight 1 
fdbe:e98b:173f:4f12::/64 dev enp0s3 proto ra metric 100 pref medium
fdcd:7fc0:8284::/64 proto ra metric 100 pref medium
    nexthop via fe80::46c:2041:d140:dae0 dev enp0s3 weight 1 
    nexthop via fe80::c24:8ac1:ca8d:c813 dev enp0s3 weight 1 
    nexthop via fe80::184b:fec6:8d00:a266 dev enp0s3 weight 1 
    nexthop via fe80::80f:73cb:ae8:b858 dev enp0s3 weight 1 
    nexthop via fe80::14a2:7660:694b:73ce dev enp0s3 weight 1 
```

So take away from this is:

* `CONFIG_IPV6_ROUTE_INFO` / `accept_ra_rt_info_max_plen` is needed for non-HAOS use cases where NetworkManager is not present, but on HAOS NM can make this work without kernel help.
* Link local ipv6 alone is not enough for thread, as NM won't be processing the RI. (This is not a problem, just need to be mindful that seeing link local addresses is not enough to expect routes to be propagated).
* The `ip` command from Alpine's busybox is *not* sufficient for debugging thread connectivity issues, as it can't handle multiple border routers.

**Reply by agners** [reply]
> Setting the sysctl is also mentioned in the OpenThread docs.

Yes, in the [Thread Border Router - Bidirectional IPv6 Connectivity and DNS-Based Service Discovery](https://openthread.io/codelabs/openthread-border-router). However, as by my findings in https://github.com/home-assistant/operating-system/discussions/2333#discussioncomment-4831967, it seems that at least `accept_ra` is essentially reverted by NetworkManager anyways. :sweat_smile: 

It seems that NetworkManager also has a way to push routes into a specific routing table (`ipv6.route-table` setting). I guess that requires them (and maybe other features) to process them in user space.

I agree with your take aways. I'd also like to add full `iproute2` to HAOS.

**Comment by agners** [comment]
I was wondering if NetworkManager's IPv6 link-local address configuration `ipv6.method=link-local` enables IPv6 Neighbor Discovery protocol. Method `link-local` is the default when "disabled" is chosen as IPv6 configuration in the Supervisor/frontend, see https://github.com/home-assistant/supervisor/pull/3829.

By experimentation it seems that NetworkManager does not process IPv6 Neighbor Discovery packets.

And looking into NetworkManager source code supports that: `nm_ndisc_start` is only called in [src/core/devices/nm-device.c#L11053](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.34.0/src/core/devices/nm-device.c#L11053), which in turn is called when method is `shared` or `auto` (see [src/core/devices/nm-device.c#L10314-10326](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.34.0/src/core/devices/nm-device.c#L10314-10326)).

It also seems that the kernels routing advertising processing is disabled by the NetworkManager in [src/core/devices/nm-device.c#L11446](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.34.0/src/core/devices/nm-device.c#L11446).

**Comment by Jc2k** [comment]
So since our last discussion I have been working with a number of users in the community to try to find the cause of instability with their Thread setups.

One case that I've been looking into is a user with"ghost" routes. When I first helped them `ip -6 route` was showing 12 routes to their Thread network - as in, it looked like they had 12 border routers. But they only had 2. And their zeroconf data confirmed that they only had 2 border routers.

Whilst the exact cause of that is not known for them, I have now been able to force my own mesh into the same state. And it is fairly trivial.

## Working setup (Debian + HA Container)

First of all, here is my production Home Assistant. This environment is showing what I consider to be expected. It's a Debian machine running containers. It is connected to my iot vlan. `accept_ra_rt_info_max_plen ` is set to 64. And it has `CONFIG_IPV6_ROUTE_INFO`. At the start of the experiment it can see 2 HomePod's:

```
/config # ip -6 route
...
fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev vlan101 proto ra metric 1024 expires 1645sec pref medium
fd59:86c6:e5a5::/64 via fe80::801:31c6:bb2:30b9 dev vlan101 proto ra metric 1024 expires 1704sec pref medium
...
```

Both of these have expiry data visible. At this point I unplugged one of the HomePod's. I left it alone for 30 minutes and when I came back:

```
/config # ip -6 route
...
fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev vlan101 proto ra metric 1024 expires 1794sec pref medium
...
```

As expected, one route had expired. The other had continued to see new RA's so the expires time has actually gone up since I last checked.

I checked `ip neigh`:

```
/config # ip -6 neigh | grep fe80::801:31c6:bb2:30b9
/config # ip -6 neigh | grep fe80::10ed:6406:4ee9:85e5 
fe80::10ed:6406:4ee9:85e5 dev vlan101 lladdr 04:99:b9:64:1b:eb router STALE 
```

As expected, the border router I unplugged has been removed.

And pinging my accessories works:

```
PING fd59:86c6:e5a5:0:a675:1612:133:abd1 (fd59:86c6:e5a5:0:a675:1612:133:abd1): 56 data bytes
64 bytes from fd59:86c6:e5a5:0:a675:1612:133:abd1: seq=0 ttl=63 time=29.049 ms
```

And for completeness:

```
ip route get  fd59:86c6:e5a5:0:a675:1612:133:abd1
fd59:86c6:e5a5:0:a675:1612:133:abd1 from :: via fe80::10ed:6406:4ee9:85e5 dev vlan101 proto ra src fd4a:2c09:9405:3ffa:65:aff:fec0:aa0a metric 1024 pref medium
```

## System with Ghosts (HAOS)

Now here is a HAOS 9.5 system. Based on the conversation in this thread, we know that NetworkManager is in charge of processing RA's.

We start off in the same state as the Debian system:

```
/config # ip -6 route
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
        nexthop via fe80::801:31c6:bb2:30b9 dev eth0 weight 1 
        nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
```

However it's worth calling out at this stage how differently formatted the output is, despite using the same version of `Ip` in both environments (I used the `iproute2` app, and I ran it inside the HA container so it's not a Debian vs build root thing). It's also interesting that the `expires` field is not present at all.

After half an hour has passed I returned and checked. Whilst the Debian system no longer had the stale route, HAOS does:

```
/config # ip -6 route
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
        nexthop via fe80::801:31c6:bb2:30b9 dev eth0 weight 1 
        nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
```

`ip neigh` knows something is wrong, but both border routers are still listed:

```
/config # ip -6 neigh | grep fe80::801:31c6:bb2:30b9
fe80::801:31c6:bb2:30b9 dev eth0 router FAILED 
/config # ip -6 neigh | grep fe80::10ed:6406:4ee9:85e5
fe80::10ed:6406:4ee9:85e5 dev eth0 lladdr 04:99:b9:64:1b:eb router DELAY 
```

My accessory cannot be pinged:

```
/config # ping6 fd59:86c6:e5a5:0:a675:1612:133:abd1
PING fd59:86c6:e5a5:0:a675:1612:133:abd1 (fd59:86c6:e5a5:0:a675:1612:133:abd1): 56 data bytes
^C
--- fd59:86c6:e5a5:0:a675:1612:133:abd1 ping statistics ---
4 packets transmitted, 0 packets received, 100% packet loss
```

Confusingly `ip route get` makes a good choice:

```
/config # ip route get  fd59:86c6:e5a5:0:a675:1612:133:abd1
fd59:86c6:e5a5:0:a675:1612:133:abd1 from :: via fe80::10ed:6406:4ee9:85e5 dev eth0 proto ra src fd4a:2c09:9405:3ffa:4629:f3e0:175e:601d metric 100 pref medium
```

As soon as we delete the stale route:

```
ip route del fd59:86c6:e5a5::/64 via fe80::801:31c6:bb2:30b9
```

ping works:

```
/config # ping6 fd59:86c6:e5a5:0:a675:1612:133:abd1
PING fd59:86c6:e5a5:0:a675:1612:133:abd1 (fd59:86c6:e5a5:0:a675:1612:133:abd1): 56 data bytes
64 bytes from fd59:86c6:e5a5:0:a675:1612:133:abd1: seq=0 ttl=63 time=28.621 ms
64 bytes from fd59:86c6:e5a5:0:a675:1612:133:abd1: seq=1 ttl=63 time=26.838 ms
```

**Reply by Jc2k** [reply]
2 initial theories:

* NM does not set any expiry and does not process expiry itself
* NM disables whichever kernel feature handles expiry (perhaps turning off `accept_ra` has that side effect?)

Looking at NM, it certainly looks like it tries to set expiry (https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.34.0/src/core/ndisc/nm-lndp-ndisc.c#L240-265). So why can't `ip` see it?

**Reply by Jc2k** [reply]
Some example `ip` incantations and the impact on the route table.

I can't seem to set different expires on different border routers:

```
ip route add fd59:86c6:e5a5::/64  dev eth0 nexthop via fe80::10ed:6406:4ee9:85e5 expires 3000 nexthop via fe80::801:31c6:bb2:30b9 expires 3000
Error: "nexthop" or end of line is expected instead of "expires"
```

But I can for the route as a whole:

```
# ip route add fd59:86c6:e5a5::/64  expires 3000 dev eth0 nexthop via fe80::10ed:6406:4ee9:85e5 nexthop via fe80::801:31c6:bb2:30b9
# ip -6 route
...
fd59:86c6:e5a5::/64 metric 1024 expires 2998sec pref medium
        nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
        nexthop via fe80::801:31c6:bb2:30b9 dev eth0 weight 1
...
```

This isn't very useful, though.

I can't create multiple routes `naively`... e.g.

```
/config # ip route add fd59:86c6:e5a5::/64  expires 3000 dev eth0 via fe80::10ed:6406:4ee9:85e5
/config # ip route add fd59:86c6:e5a5::/64  expires 3000 dev eth0 via fe80::801:31c6:bb2:30b9
RTNETLINK answers: File exists
```

I was able to verify that if i use the `nexthop` scheme on my Debian box (no NetworkManager) i can get HAOS style output:

```
# ip -6 route
::1 dev lo proto kernel metric 256 pref medium
fd59:86c6:e5a4::/64 metric 1024 expires 2995sec pref medium
	nexthop via fe80::10ed:6406:4ee9:85e5 dev vlan101 weight 1 
	nexthop via fe80::801:31c6:bb2:30b9 dev vlan101 weight 1 
````

So i'm now increasingly thinking that NM is creating the routes differently to what the kernel does for option 24, and the mechanism that it uses isn't able to do per gateway expires.

**Reply by Jc2k** [reply]
`ip monitor route` for both environments. SSH'd into both, and monitor commands running side by side. Starting with 1 border router on, then turning on 2nd border router. 

Debian, no NM:

```
# ip route monitor
fd59:86c6:e5a5::/64 via fe80::c95:1bc1:972f:7450 dev vlan101 proto ra metric 1024 pref medium
```

For completeness, the resulting RT:

```
fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev vlan101 proto ra metric 1024 expires 1625sec pref medium
fd59:86c6:e5a5::/64 via fe80::c95:1bc1:972f:7450 dev vlan101 proto ra metric 1024 expires 1777sec pref medium
```

HAOS / NM

```
# ip route monitor
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
	nexthop via fe80::c95:1bc1:972f:7450 dev eth0 weight 1 
	nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
Deleted fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev eth0 proto ra metric 100 pref medium
```

And the new route table:

```
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
	nexthop via fe80::c95:1bc1:972f:7450 dev eth0 weight 1 
	nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
```

**Reply by Jc2k** [reply]
It looks like NM and the kernel get out of sync. Right now I see this:

```
Feb 23 13:41:17 homeassistant NetworkManager[194442]: <debug> [1677159677.4427] platform: (eth0) route-sync: adding route type unicast fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev 2 metric 100 mss 0 rt-src ndisc failed with EEXIST, however we cannot find such a route
```

But I see this:

```
/config # ip -6 route
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
	nexthop via fe80::464:2037:b48b:b569 dev eth0 weight 1 
	nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
```

I should add that right now `nmcli --fields all d show eth0` gives:

```
IP6.ROUTE[1]:                           dst = fd69:3e50:e165:4bf8::/64, nh = ::, mt = 100
IP6.ROUTE[2]:                           dst = fe80::/64, nh = ::, mt = 100
```

If I look at the logs for when I turned the 2nd HomePod on I see this:

```
Feb 23 13:13:40 homeassistant NetworkManager[194442]: <debug> [1677158020.5536] platform-linux: do-add-ip6-address[2: fd69:3e50:e165:4bf8:32ab:b6b4:2690:a16f]: success
Feb 23 13:13:40 homeassistant NetworkManager[194442]: <debug> [1677158020.5538] platform: (eth0) route: append     IPv6 route: type unicast fd59:86c6:e5a5::/64 via fe80::464:2037:b48b:b569 dev 2 metric 100 mss 0 rt-src ndisc
Feb 23 13:13:40 homeassistant NetworkManager[194442]: <debug> [1677158020.5541] platform-linux: do-add-ip6-route[type unicast fd59:86c6:e5a5::/64 via fe80::464:2037:b48b:b569 dev 2 metric 100 mss 0 rt-src rt-ra]: success
Feb 23 13:13:40 homeassistant NetworkManager[194442]: <debug> [1677158020.5542] platform: (eth0) ip6-route: delete type unicast fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev 2 metric 100 mss 0 rt-src rt-ra is-external
Feb 23 13:13:40 homeassistant NetworkManager[194442]: <debug> [1677158020.5544] platform: (eth0) signal: route   6 removed: type unicast fd59:86c6:e5a5::/64 via fe80::10ed:6406:4ee9:85e5 dev 2 metric 100 mss 0 rt-src rt-ra is-external
```

It adds a route via b569 (office homepod) and tries to delete 85e5 (bedroom HomePod). We don't get any more RA's from b569 (cos I turned it back off) and now 85e5 is still sending RA's but they are hitting the EEXIST error.

Then I turned the office one back on.

```
fd59:86c6:e5a5::/64 proto ra metric 100 pref medium
	nexthop via fe80::464:2037:b48b:b569 dev eth0 weight 1 
	nexthop via fe80::10ed:6406:4ee9:85e5 dev eth0 weight 1 
	nexthop via fe80::2c:659f:47da:9226 dev eth0 weight 1 
```

Now 9226 is a new route so it can be added:

```
Feb 23 13:51:08 homeassistant NetworkManager[194442]: <debug> [1677160268.9449] platform: (eth0) route: append     IPv6 route: type unicast fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 dev 2 metric 100 mss 0 rt-src ndisc
Feb 23 13:51:08 homeassistant NetworkManager[194442]: <debug> [1677160268.9452] platform-linux: do-add-ip6-route[type unicast fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 dev 2 metric 100 mss 0 rt-src rt-ra]: success
Feb 23 13:51:23 homeassistant NetworkManager[194442]: <debug> [1677160283.9225] ndisc[0x5575e3a850,"eth0"]:   route fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 pref medium exp 1800.000
```

But it is immediately invisible:

```
Feb 23 13:51:23 homeassistant NetworkManager[194442]: <debug> [1677160283.9244] platform: (eth0) route: append     IPv6 route: type unicast fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 dev 2 metric 100 mss 0 rt-src ndisc
Feb 23 13:51:23 homeassistant NetworkManager[194442]: <debug> [1677160283.9248] platform-linux: do-add-ip6-route[type unicast fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 dev 2 metric 100 mss 0 rt-src rt-ra]: failure 17 (File exists)
Feb 23 13:51:23 homeassistant NetworkManager[194442]: <debug> [1677160283.9249] platform: (eth0) route-sync: adding route type unicast fd59:86c6:e5a5::/64 via fe80::2c:659f:47da:9226 dev 2 metric 100 mss 0 rt-src ndisc failed with EEXIST, however we cannot find such a route
```

I am wondering if there might be 2 problems:

* When the 2nd route gets added, the kernel converts the route into a single route with multiple next hops (as above). When this happens, NM can no longer see the route (hence the error above)
* Why did it try to delete a route that hadn't expired?

**Reply by Jc2k** [reply]
Sounds like this is fixed in or before 1.40 as Debian doesn't hit this. In fact, that error doesn't even exist any more. So the buildroot update you mentioned might fix "ghost routes"

However, the new behaviour is not that great. It looks like the last RA wins. Only a single route at any time. So for my user with 8 HomePods, where the RA's go out every 3 minutes, his default route to the mesh is going to churn 2-3 times a minute.

**Reply by agners** [reply]
Further investigations with @Jc2k brought new facts to daylight. Test bed was again Apple BR devices.

With OS 10.0.rc1 and NetworkManager 1.40.10, the behavior is different: Only a single route gets added, each time a new RIO comes in the route of the previous BR gets overwritten. This effectively leads to the routing table to flip between two states:

```
fda8:1199:510f::/64 via fe80::830:4a86:36f0:5eb5 dev eth0 proto ra metric 100 pref medium
```

```
fda8:1199:510f::/64 via fe80::5f:ce7c:bdf:d7dc dev eth0 proto ra metric 100 pref medium
```

Router Advertisements are sent every 180s, so on average the routes flip every 90s. It is unclear if this is problematic, for a simple ping it doesn't seem to matter. However, if one of the routers fails, it can take up to 180s for the others advertisement to re-add itself and therefor "fix" the routing table.

The following BRs are used:

- `fe80::830:4a86:36f0:5eb5` is my Apple TV 4K BR
- `fe80::5f:ce7c:bdf:d7dc` is my Apple HomePod, which I will take offline later

The reason is that only a single route to a given destination network with a given prefix length is used. Each subsequent RIO is updating that same route, see:
https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.40.16/src/core/ndisc/nm-ndisc.c#L656

Adding a check to match the gateway in this if statement changes NetworkManger's behavior such that it creates two routes:

```
fda8:1199:510f::/64 proto ra metric 100 pref medium
        nexthop via fe80::830:4a86:36f0:5eb5 dev enp1s0 weight 1
        nexthop via fe80::5f:ce7c:bdf:d7dc dev enp1s0 weight 1
```

Note that the routes don't carry an `expires` attribute. It seems that NetworkManger is handling this in user space. The default route lifetime is 1800s (or 30min).

However, with two routes now, whichever route is active, will be the one which is chosen (in my testing it was always the lower). Which one it is can also be checked with `ip route get fda8:1199:510f::...`

When I now unplug the Apple HomePod device (breaking the lower route), the Thread device is no longer reachable. In practice, this leads to not working routes between 27 and 30min (depending on when the RIO came in, until the bad route is purged by NetworkManager)!

A little Ubuntu 22.04.1 LTS Server Side quest, open up details if interested.

<details>
On a Ubuntu 22.04.1 LTS Server systemd-networkd 249 is running. It seems that this device is suffering a similar problem. However, in this case the routes are grouped. Presumably they then also expire together (which might be a problem of systemd-networkd):

```
fda8:1199:510f::/64 proto ra metric 1024 expires 1778sec pref medium
        nexthop via fe80::830:4a86:36f0:5eb5 dev eth0 weight 1 
        nexthop via fe80::5f:ce7c:bdf:d7dc dev eth0 weight 1 
```
</details>

Shouldn't Linux detect non-working routes somehow?

@Jc2k pointed me to `CONFIG_IPV6_ROUTER_PREF` which if enabled [checks routes](https://github.com/torvalds/linux/blob/3a93e40326c8f470e71d20b4c42d36767450f38f/net/ipv6/route.c#L693-L698) and should influence route selection. HAOS 10.0.rc1 does not have that flag enabled currently. However, even after enabling it things did not work. It turned out that globally enabled IPv6 routing support `net.ipv6.conf.all.forwarding` ("router mode") influences behavior as well (see [net/ipv6/route.c](https://github.com/torvalds/linux/blob/3a93e40326c8f470e71d20b4c42d36767450f38f/net/ipv6/route.c#L2213-L2214)). The has been implemented back in [2006 with this commit](https://github.com/torvalds/linux/commit/ea659e0775281e1c02556e939bf749bb4f55e50c), and it seems that RFC4191 makes a difference between a router and a host device.

With routing disabled, the route availability check seems to work flawlessly! Route selection is almost instant, at least using ping no big hick-up can be detected.

Note `ip route get` just seems to inspect the routing table and does not take reach-ability into account! The kernel seems to select a different route in the end. E.g. in this example `fe80::5f:ce7c:bdf:d7dc` is unavailable, pinging to a Thread device works, but  `ip route get` still returns that unavailable route.

```
# ip route get fda8:1199:510f:0:c013:d6d5:135c:fdeb
fda8:1199:510f:0:c013:d6d5:135c:fdeb from :: via fe80::5f:ce7c:bdf:d7dc dev eth0 proto ra src fd14:949b:c9cc:10::c8a metric 100 pref medium
# ping fe80::5f:ce7c:bdf:d7dc%eth0
No response from fe80::5f:ce7c:bdf:d7dc%eth0
# ping fda8:1199:510f:0:c013:d6d5:135c:fdeb
fda8:1199:510f:0:c013:d6d5:135c:fdeb is alive!
```

With these findings, it is probably advisable to:
- Enable at least `CONFIG_IPV6_ROUTER_PREF`
- Disable global IPv6 routing again (`net.ipv6.conf.all.forwarding=0`)

**Reply by agners** [reply]
Instead of globally disable IPv6 routing again I came up with a patch to force enable router availability probing. I've also sent an [RFC PATCH to the mailing list](https://lore.kernel.org/netdev/b9182b02829b158d55acc53a0bcec1ed667b2668.1680000784.git.stefan@agner.ch/T/#u), maybe this is something which Linux can support upstream.

As for the NetworkManager change: It seems to work reliable, and I suggested the change in the NetworkManager issue previously opened to discuss issues we've been seen: https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1232#note_1841630.

Depending on outcome of that discussion, I'll try to upstream this change as well.

PR https://github.com/home-assistant/operating-system/pull/2434 applies the two changes to HAOS.

**Comment by MattWestb** [comment]
I have one similar problem with my 3 OTBR on 3 machines.
- NanoPI Neo with Armbian Ubuntu 32 and supervised HA minimal Zigbee test and home kit for Dirigera.
- RK3228A with Ambian Ubuntu 32 and supervised HA medium Zigbee test and active OTBR and home kit thread devices.
- RK3318 with Armbian Ubuntu 64 and supervised HA large (30 + devices) Zigbee test with matter server and Hyper HDR.

Other thread devices:
- 2 EVE Energy thread running the last home kit firmware and not the matter upgrade.
- Nanoleaf RGBW Light current thread / home kit firmware.
- ERF32MG21 OTFD-CLI device (Running Silabs OTFD with CLI so can manage the device and network with CLI).


With HA adding possibility for home kit controller Provision devices from BT to thread i was getting 2 EVE Eanergy and one Nanoleaf RGBW light joining the tread network with the 3 OTBR and one OTFD-CLI device but HA is having problem keeping contact to them.
In the log is looks like mDNS is fighting having the default name of the addon but that  is fixed yesterdays addon fix (1.0.1). Still in the log is address conflicts then mDNS is trying allocating routers for the same device in multiple OTBRs and is braking the connection to the thread devices.
The funny thing is looks like the OTBR web GUI is working OK and can manager the network and looking on the topography in it.

By disabling the OTBR and only running the Zigbeed on the first ans last host the second one is working great and its fast as one flash then sending and receiving command to and from my tread device !!!

If the multi OTBR was OK my thread network have being moved to my production system but is not critical devices on it so its staying until its working OK.

If its something i can helping doing diagnostic test and patches i one good setup for doing the test but dont have very deep knowledge in IPV6 and Linux but can doing it with little help.

And one great thanks you both for getting all thins working great so fare with all work being done !!
Now its only the small pieces that need being fixed and if i can helping pleas hit my up.
